vmm/src/device_manager.rs

```diff
@@ -1007,21 +1018,38 @@ impl DeviceManager {
                 1
             };
 
-        let start_of_device_area = memory_manager.lock().unwrap().start_of_device_area().0;
-        let end_of_device_area = memory_manager.lock().unwrap().end_of_device_area().0;
+        let start_of_mmio32_area = layout::MEM_32BIT_DEVICES_START.0;
+        let end_of_mmio32_area = layout::MEM_32BIT_DEVICES_START.0 + layout::MEM_32BIT_DEVICES_SIZE;
+
+        // Start each PCI segment mmio32 range on a page-size boundary
+        let pci_segment_mmio32_size = (end_of_mmio32_area - start_of_mmio32_area + 1)
+            / ((4 << 10) * num_pci_segments as u64)
+            * (4 << 10);
 
-        // Start each PCI segment range on a 4GiB boundary
-        let pci_segment_size = (end_of_device_area - start_of_device_area + 1)
+        let mut pci_mmio32_allocators = vec![];
+        for i in 0..num_pci_segments as u64 {
+            let mmio_start = start_of_mmio32_area + i * pci_segment_mmio32_size;
+            let allocator = Arc::new(Mutex::new(
+                AddressAllocator::new(GuestAddress(mmio_start), pci_segment_mmio32_size).unwrap(),
+            ));
+            pci_mmio32_allocators.push(allocator)
+        }
+
+        let start_of_mmio64_area = memory_manager.lock().unwrap().start_of_device_area().0;
+        let end_of_mmio64_area = memory_manager.lock().unwrap().end_of_device_area().0;
+
+        // Start each PCI segment mmio64 range on a 4GiB boundary
+        let pci_segment_mmio64_size = (end_of_mmio64_area - start_of_mmio64_area + 1)
             / ((4 << 30) * num_pci_segments as u64)
             * (4 << 30);
 
-        let mut pci_mmio_allocators = vec![];
+        let mut pci_mmio64_allocators = vec![];
         for i in 0..num_pci_segments as u64 {
-            let mmio_start = start_of_device_area + i * pci_segment_size;
+            let mmio_start = start_of_mmio64_area + i * pci_segment_mmio64_size;
             let allocator = Arc::new(Mutex::new(
-                AddressAllocator::new(GuestAddress(mmio_start), pci_segment_size).unwrap(),
+                AddressAllocator::new(GuestAddress(mmio_start), pci_segment_mmio64_size).unwrap(),
             ));
-            pci_mmio_allocators.push(allocator)
+            pci_mmio64_allocators.push(allocator)
```

Comment:
I believe it is cleaner to use a "closure" like below: 

```
let create_allocators = |start, end, num_pci_segments: u16, aligment: u64| {};
```

Reply:
Ack. Love it. A lot less repetitive. Updated.